### PR TITLE
ARROW-7428: [Format][C++] Add serialization for CSF sparse tensors

### PR DESCRIPTION
**Pull request by rok**
This is to solve [ARROW-7428](https://issues.apache.org/jira/browse/ARROW-7428).

### CODE REVIEW

**Comment by rok** [comment]
@pitrou @mrkn - I'm tying to serialize [vectors of Buffers](https://github.com/apache/arrow/pull/6340/files#diff-c07b6024c7170d1dcf02b239f3bfb8b6R160) but I'm getting [wrong buffer offsets and lengths](https://github.com/apache/arrow/pull/6340/checks?check_run_id=478364955#step:5:1803) on deserializing. Any idea why this is? I'm a bit stuck.

**Comment by mrkn** [comment]
@rok I've begun to look it.

File: cpp/src/arrow/ipc/metadata_internal.cc [hunk 1]
```diff
@@ -961,6 +961,58 @@ Status MakeSparseMatrixIndexCSX(FBB& fbb, const SparseIndexType& sparse_index,
...
+  auto indices_type_offset = flatbuf::CreateInt(fbb, indices_value_type.bit_width(),
+                                                indices_value_type.is_signed());
+
+  std::vector<const flatbuf::Buffer*> indptr;
+  std::vector<const flatbuf::Buffer*> indices;
+
+  for (int i = 0; i < ndim - 1; ++i) {
+    const BufferMetadata& buffer = buffers[i];
+    flatbuf::Buffer tmp_buffer(buffer.offset, buffer.length);
+    indptr.emplace_back(&tmp_buffer);
```
Comment (mrkn) [inline cpp/src/arrow/ipc/metadata_internal.cc]: `tmp_buffer` is a temporary variable, so `&tmp_buffer` is invalid just after exiting this block.

File: cpp/src/arrow/ipc/metadata_internal.cc [hunk 2]
```diff
@@ -961,6 +961,58 @@ Status MakeSparseMatrixIndexCSX(FBB& fbb, const SparseIndexType& sparse_index,
...
+  }
+
+  for (int i = ndim - 1; i < 2 * ndim - 1; ++i) {
+    const BufferMetadata& buffer = buffers[i];
+    flatbuf::Buffer tmp_buffer(buffer.offset, buffer.length);
+    indices.emplace_back(&tmp_buffer);
```
Comment (mrkn) [inline cpp/src/arrow/ipc/metadata_internal.cc]: same as above

File: cpp/src/arrow/ipc/reader.cc [hunk 3]
```diff
@@ -954,6 +955,36 @@ Result<std::shared_ptr<SparseIndex>> ReadSparseCSXIndex(
...
+  std::vector<int64_t> indices_size, axis_order;
+
+  RETURN_NOT_OK(internal::GetSparseCSFIndexMetadata(
+      sparse_index, &indices_size, &axis_order, &indptr_type, &indices_type));
+
+  for (int64_t i = 0; i < ndim - 1; ++i) {
+    ARROW_ASSIGN_OR_RAISE(indptr_data[i], file->ReadAt(indptr_buffers->Get(i)->offset(),
+                                                       indptr_buffers->Get(i)->length()));
+  }
+  for (int64_t i = ndim - 1; i < 2 * ndim - 1; ++i) {
```
Comment (mrkn) [inline cpp/src/arrow/ipc/reader.cc]: `i` should start from `0` for `indices_data` and `indices_buffers`.

**Review by mrkn — CHANGES_REQUESTED** [review]
I haven't tried the pull-request yet but left some comments.

File: cpp/src/arrow/ipc/metadata_internal.cc (hunk 1 above)
Comment (rok) [reply cpp/src/arrow/ipc/metadata_internal.cc]: Changed. The objects now stay in memory after the for loops have completed.

File: cpp/src/arrow/ipc/reader.cc (hunk 3 above)
Comment (rok) [reply cpp/src/arrow/ipc/reader.cc]: Changed.
It seems the offset and length values here are either not correctly read from metadata or something goes wrong at write-time. Either way I'm getting:
```
'_error_or_value26.status()' failed with IOError: Read out of bounds (offset = 94719482752432, size = 1688849860263936) in file of size 264

```

File: cpp/src/arrow/ipc/writer.cc [hunk 4]
```diff
@@ -810,6 +815,25 @@ class SparseTensorSerializer {
...
+    std::vector<std::shared_ptr<Buffer>> indptr_buffers, indices_buffers;
+
+    for (const std::shared_ptr<arrow::Tensor> indptr : sparse_index.indptr()) {
+      indptr_buffers.emplace_back(indptr->data());
+    }
+    for (const std::shared_ptr<arrow::Tensor> indices : sparse_index.indices()) {
+      indices_buffers.emplace_back(indices->data());
+    }
+
+    RETURN_NOT_OK(ConcatenateBuffers(indptr_buffers, default_memory_pool(), &indptrs));
```
Comment (rok) [inline cpp/src/arrow/ipc/writer.cc]: @mrkn @pitrou do you know of a better approach to do this given the [schema](https://github.com/apache/arrow/blob/7e325a24c25575f83095afe84413beda4e05a64a/format/SparseTensor.fbs#L160)?

Comment (pitrou) [reply cpp/src/arrow/ipc/writer.cc]: I'm not sure I understand. `SparseTensorIndexCSF.indptrBuffers` and `SparseTensorIndexCSF.indicesBuffers` are supposed to be arrays of buffers. Why do you need to concatenate them?

Comment (rok) [reply cpp/src/arrow/ipc/writer.cc]: My understanding is that [`Assemble`](https://github.com/apache/arrow/blob/7e325a24c25575f83095afe84413beda4e05a64a/cpp/src/arrow/ipc/writer.cc#L774) expects a vector of buffers to serialize.
Alternatively we could do:
```
    for (const std::shared_ptr<arrow::Tensor> indptr : sparse_index.indptr()) {
      out_->body_buffers.emplace_back(indptr->data());
    }
    for (const std::shared_ptr<arrow::Tensor> indices : sparse_index.indices()) {
      out_->body_buffers.emplace_back(indices->data());
    }
```

Comment (pitrou) [reply cpp/src/arrow/ipc/writer.cc]: That would look better, but you also have to fill the `SparseTensorIndexCSF` fields appropriately.

By the way, I think you should tackle Python serialization in a separate PR and focus on IPC here.

Comment (rok) [reply cpp/src/arrow/ipc/writer.cc]: > By the way, I think you should tackle Python serialization in a separate PR and focus on IPC here.

Will do.

**Comment by rok** [comment]
I think this is ready for review.

**Comment by rok** [comment]
Ping.

**Review by wesm — COMMENTED** [review]
As far as the code, this looks fine to me. If someone who knows more about sparse matrices / tensors could have a look at the Format changes, that would be helpful before I merge this

File: cpp/src/arrow/ipc/metadata_internal.cc [hunk 5]
```diff
@@ -957,6 +957,67 @@ Status MakeSparseMatrixIndexCSX(FBB& fbb, const SparseIndexType& sparse_index,
...
+    indices.push_back({indices_offset, size});
+    indices_offset += size + padding;
+  }
+
+  auto fb_indices = fbb.CreateVectorOfStructs(indices);
+  auto fb_indptr = fbb.CreateVectorOfStructs(indptr);
+
+  std::vector<int> axis_order;
+  for (int i = 0; i < ndim; ++i) {
+    axis_order.emplace_back(static_cast<const int>(sparse_index.axis_order()[i]));
```
Comment (mrkn) [inline cpp/src/arrow/ipc/metadata_internal.cc]: Why you cast the value to `const int` instead of just `int`?

**Review by mrkn — CHANGES_REQUESTED** [review]
I leave one comment.

File: cpp/src/arrow/ipc/read_write_test.cc [hunk 6]
```diff
@@ -1289,6 +1289,66 @@ class TestSparseTensorRoundTrip : public ::testing::Test, public IpcTestFixture
...
+                                                     sparse_index.indptr()[i]->size());
+    }
+    for (int64_t i = 0; i < ndim; ++i) {
+      indices_length += BitUtil::RoundUpToMultipleOf8(index_elem_size *
+                                                      sparse_index.indices()[i]->size());
+    }
+    const int64_t data_length =
+        BitUtil::RoundUpToMultipleOf8(elem_size * sparse_tensor.non_zero_length());
+    const int64_t expected_body_length = indptr_length + indices_length + data_length;
+    ASSERT_EQ(expected_body_length, body_length);
```
[the remark this answers is not in the document]
Comment (rok) [reply cpp/src/arrow/ipc/read_write_test.cc]: Not really. I kept it to stay consistent with `CheckSparseCOOTensorRoundTrip` and `CheckSparseCSXMatrixRoundTrip`.

File: cpp/src/arrow/ipc/metadata_internal.cc [hunk 7]
```diff
@@ -957,6 +957,67 @@ Status MakeSparseMatrixIndexCSX(FBB& fbb, const SparseIndexType& sparse_index,
...
+  }
+  indices_offset = indptr_offset;
+  for (const std::shared_ptr<arrow::Tensor> tensor : sparse_index.indices()) {
+    const int64_t size = tensor->data()->size() / indices_elem_size;
+    const int64_t padding =
+        BitUtil::RoundUpToMultipleOf8(size * indices_elem_size) - size;
+
+    indices.push_back({indices_offset, size});
+    indices_offset += size + padding;
+  }
```
[the remark this answers is not in the document]
Comment (rok) [reply cpp/src/arrow/ipc/metadata_internal.cc]: Switching to `kArrowIpcAlignment` for padding. Is that what you had in mind?

File: cpp/src/arrow/ipc/metadata_internal.cc (hunk 5 above)
Comment (rok) [reply cpp/src/arrow/ipc/metadata_internal.cc]: Changed.

**Comment by rok** [comment]
Thanks for the review @wesm and @mrkn. I've pushed another commit.

File: cpp/src/arrow/ipc/metadata_internal.cc (hunk 7 above)
[the remark this answers is not in the document]
Comment (wesm) [reply cpp/src/arrow/ipc/metadata_internal.cc:975]: Some other functions have an alignment option/parameter. It's not a big deal just was thinking out loud about it

**Review by wesm — APPROVED** [review]
+1. Merging this. Thanks all

**Comment by rok** [comment]
Thanks @wes @mrkn and @pitrou!

**Pull request closed without merging**